src/Akka.Cluster.Hosting/AkkaClusterHostingExtensions.cs

```diff
@@ -179,9 +177,44 @@ public static class AkkaClusterHostingExtensions
                         .WithRememberEntities(shardOptions.RememberEntities)
                         .WithStateStoreMode(shardOptions.StateStoreMode), extractEntityId, extractShardId);
 
-                // TODO: should throw here if duplicate key used
+                registry.Register<TKey>(shardRegion);
+            });
+        }
+        
+        public static AkkaConfigurationBuilder WithShardRegion<TKey>(this AkkaConfigurationBuilder builder,
+            string typeName,
+            Func<ActorSystem, IActorRegistry, Func<string, Props>> compositePropsFactory, IMessageExtractor messageExtractor, ShardOptions shardOptions)
```

Comment:
We take a ` Func<ActorSystem, IActorRegistry, Func<string, Props>>` here now, which allows for methods like this to be created when a `ShardRegion` is starting:

```csharp
 .WithShardRegion<MyEntityActor>("entities", (system, registry) =>
                {
                    var tLevel = registry.Get<MyTopLevelActor>();
                    return s => Props.Create(() => new MyEntityActor(s, tLevel));
                }, new Extractor(), new ShardOptions() { Role = "my-host", StateStoreMode = StateStoreMode.DData });
```

Reply:
Should we make `ExtractEntityId` and `ExtractShardId` obsolete in favor of `IMessageExtractor` in 1.5?

Reply:
Nevermind, should've read the rest of the changes first